### PR TITLE
feat: benchmark reading many small chunks

### PR DESCRIPTION
**Pull request by coryan**
New benchmark, verifies we can achieve high bandwidth when downloading
many small chunks.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/3313)

### CODE REVIEW

File: google/cloud/storage/benchmarks/run_benchmarks_testbench.sh [hunk 1]
```diff
@@ -76,6 +87,7 @@ run_example ./storage_throughput_vs_cpu_benchmark \
       --maximum-chunk-size=32KiB \
       --duration=1s
 
+
```
Comment (devbww) [inline google/cloud/storage/benchmarks/run_benchmarks_testbench.sh]: Remove added extra blank line?

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+    std::cout << usage << "\n";
+  }
+
+  if (wants_description) {
+    std::cout << kDescription << "\n";
+  }
+
+  if (unparsed.size() > 2) {
+    std::ostringstream os;
+    os << "Unknown arguments or options\n" << usage << "\n";
```
Comment (devbww) [inline google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: Non-actionable observation: Newlines in status messages are a little odd, particularly terminating ones.

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+                     (chunk + 1) * options.chunk_size});
+    total_bytes += options.chunk_size;
+  }
+  work_queue.Shutdown();
+  for (auto& t : workers) {
+    t.get();
+  }
+  auto const elapsed = std::chrono::steady_clock::now() - download_start;
+  return {total_bytes,
+          std::chrono::duration_cast<std::chrono::microseconds>(elapsed)};
```
Comment (devbww) [inline google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: Nit: Other places "use" these.

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+    if (group.size() >= max_group_size) {
+      tasks.emplace_back(std::async(std::launch::async, &CreateGroup, client,
+                                    bucket_name, options, std::move(group)));
+      group = {};  // after a move, must assign to guarantee it is valid.
+    }
+  }
+  if (!group.empty()) {
+    tasks.emplace_back(std::async(std::launch::async, &CreateGroup, client,
+                                  bucket_name, options, std::move(group)));
+    group = {};  // after a move, must assign to guarantee it is valid.
```
Comment (devbww) [inline google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: This one seems unnecessary, and therefore a little misleading.

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+in the test.  The bucket is deleted at the end of the test. The name of the
+bucket is selected at random, so multiple instances of this benchmark can run
+simultaneously. The bucket uses the `STANDARD` storage class, in a region set
+via the command-line.
+
+After creating this bucket the program creates a number of objects, all the
+objects have the same contents, but the contents are generated at random.
+
+The size of the objects can be configured in the command-line, but they are
+typically 250MiB is size. The objects are constructed by writing N identical
```
Comment (devbww) [inline google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: s/is size/in size/

And perhaps "typically" should be "by default"?

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+
+The number of chunks selected by the main thread is configurable from the
+command-line. After the prescribed number of chunks is generated by the main
+thread, the program shuts down the worker queue, waits for all the worker
+threads to finish, and reports the effective download throughput.
+
+Then the program removes all the objects in the bucket and reports the time
+taken to delete each one.
+
+A helper script in this directory can generate pretty graphs from the outut of
```
Comment (devbww) [inline google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: s/outut/output/

Comment (coryan) [reply google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: Fixed, thanks

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: Fixed, thanks.

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: Removed.

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: Fixed.

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: Ack.

File: google/cloud/storage/benchmarks/run_benchmarks_testbench.sh (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/benchmarks/run_benchmarks_testbench.sh]: Fixed.

**Pull request merged**